### PR TITLE
New approach for Proposal sub-graph conversion for TF OD API models

### PR DESCRIPTION
**Pull request by lazarevevgeny**
Root cause analysis: There is a customer TF OD API Faster-RCNN model which has not good enough accuracy because of the Proposal operation which is inserted with the `ObjectDetectionAPIProposalReplacement` in the MO to replace a big sub-graph of the original TF model. The investigations showed that it is possible to replace this part of the graph with the DetectionOutput operation to fix the accuracy issues.

Solution: Introduce a new parameter for the `ObjectDetectionAPIProposalReplacement` in the transformation configuration file which will define what operation to insert: Proposal or the DetectionOutput (user Proposal by default for backward compatibility). Inserting the DetectionOutput operation has a limitation related to input images which is described in the code comment for the transformation and described in the documentation. The `ObjectDetectionAPIProposalReplacement` transformation has been split into several parts to have maximum reuse between cases when we insert Proposal and DetectionOutput.

Additional info: I have run manually e2e tests for all Faster and Mask-RCNN models which we have in validation with current flow and updated flow and no regressions were identified. I will also add a dedicated e2e test to test a new flow and include it to the pre-commit (**WIP**).

Ticket: 49852

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests: N/A due to complexity of the transformations
* [x]  Framework operation tests: N/A no new ops enabled
* [x]  Transformation tests: N/A due to complexity of the transformations
* [x]  Model Optimizer IR Reader check: Done

Documentation:
* [x]  Supported frameworks operations list: N/A no new ops enabled
* [x]  Guide on how to convert the **public** model: A significant document refactoring will be done as part of another activity.
* [x]  User guide update: Done

### CODE REVIEW

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 1]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+        :param graph: the graph to operate on
+        :param match: the object containing information about the matched sub-graph
+        :param pipeline_config: object containing information from the pipeline.config file of the model
+        :return: the dictionary with mapping information needed for other transformations
+        """
+        max_proposals = _value_or_raise(match, pipeline_config, 'first_stage_max_proposals')
+
+        # Convolution/matmul node that produces classes predictions
+        # Transpose result of the tensor with classes permissions so it will be in a correct layout for Softmax
+        predictions_node = backward_bfs_for_operation(match.single_input_node(1)[0], ['Add'])[0]
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1231]: should we check a number of returned nodes here?

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 2]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+
+        ########################################################################################
+        # prepare input with class probabilities
+        reshape_classes_node = create_op_node_with_second_input(graph, Reshape, int64_array([0, -1, 2]),
+                                                                dict(name='predictions/Reshape'))
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1236]: Probably it would be nice to explain about why we needs such shape

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 3]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+        # transpose from NCHW to NHWC will be inserted as input to the Reshape automatically. This is expected
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1237]: Does it mean that it does not work if `--disable_nhwc_to_nchw` is turned on? Do we track it?

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 4]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+
+        hwhw_float = Cast(graph,
+                          {'dst_type': data_type_str_to_np(graph.graph['cmd_params'].data_type)}).create_node([hwhw])
+        scaled_anchors = Div(graph, {'name': 'scaled_anchors'}).create_node([yxyx_anchors, hwhw_float])
+
+        flattened_anchors = create_op_with_const_inputs(graph, Reshape, {1: int64_array([1, 1, -1])},
+                                                        {'name': 'flattened_anchors'}, scaled_anchors)
+        cropped_anchors = AttributedClamp(graph, {'min': 0.0, 'max': 1.0, 'name': 'clamped_xyxy',
+                                                  'nchw_layout': True}).create_node([flattened_anchors])
+        mark_as_correct_data_layout(flattened_anchors)
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1278]: It is a tricky line of code. I think it is nice to add a comment about this

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 5]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+
+        # create tensor of shape [4] with variance values which then are tiled by the number of boxes which is obtained
+        # from the 'yxyx_anchors' node
+        variances = Const(graph, {'value': np.float32([1.0 / variance_x, 1.0 / variance_y, 1.0 / variance_width,
+                                                       1.0 / variance_height])}).create_node()
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1283]: should we check that `variance_...` is non-zero?

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 6]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+
+        anchors_shape = Shape(graph, {'name': 'anchors_shape'}).create_node([yxyx_anchors])
+        anchors_count = node_to_get_shape_value_of_indices(anchors_shape, [0])
+        tiled_anchors = Tile(graph, {'name': 'tiled_anchors'}).create_node([variances, anchors_count])
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1287]: it is not a tensor of anchors. Can we rather to rename it to `tiled_variances`?

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 7]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+                              'keep_top_k': max_proposals,
+                              'normalized': True,
+                              'num_classes': 2,
+                              'objectness_score': 0,
+                              'share_location': True,
+                              'top_k': 6000,
+                              'variance_encoded_in_target': False,
+                              'nms_threshold': _value_or_raise(match, pipeline_config, 'first_stage_nms_iou_threshold'),
+                              'name': 'first_do',
+                              }).create_node([reshape_box_logits, flattened_conf, duplicate_anchors])
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1314]: I think wrong identation.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 8]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+                              'clip_after_nms': True,
+                              'clip_before_nms': False,
+                              'code_type': 'caffe.PriorBoxParameter.CENTER_SIZE',
+                              'confidence_threshold': 0.0,
+                              'decrease_label_id': False,
+                              'input_height': 1,
+                              'input_width': 1,
+                              'keep_top_k': max_proposals,
+                              'normalized': True,
+                              'num_classes': 2,
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1307]: can we have arbitrary number of classes for this? Is it valueable?

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 9]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+                              'objectness_score': 0,
+                              'share_location': True,
+                              'top_k': 6000,
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1310]: Why is it hardcoded? Can we get this number from fw?

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 10]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+                              'variance_encoded_in_target': False,
+                              'nms_threshold': _value_or_raise(match, pipeline_config, 'first_stage_nms_iou_threshold'),
+                              'name': 'first_do',
+                              }).create_node([reshape_box_logits, flattened_conf, duplicate_anchors])
+        # DetectionOutput output tensor has x and y coordinates swapped
+
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1316]: blank line. is it needed? as if something is missed.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 11]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+        # switch to 3D to avoid issues that part of the model with 4D shapes should be inferred in NCHW layout
+        do_3d = create_op_with_const_inputs(graph, Squeeze, {1: int64_array(0)}, {'name': do.name + '/squeeze_do'}, do)
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1318]: need to capitalize letters to be aligned with others: '/squeeze_do' -> '/Squeeze'

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 12]
```diff
@@ -1185,14 +1198,162 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
...
+        mark_as_correct_data_layout(do_3d)
+
+        do_split = create_op_node_with_second_input(graph, Split, int64_array(2), {'num_splits': 7, 'nchw_layout': True,
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1321]: I think it is useful to add a comment about why you split into 7 parts.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 8 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1307]: maybe it makes sense to have it as a parameter for this class because this number is meet in several places

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 5 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1291]: It is not possible that they are equal to zero.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 6 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py]: Correct. Done.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 8 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1315]: In this case number of classes is always 2: background class and foreground class. This part of the model just detects if the bounding box contains an object or not. The second DetectionOutput which is added in the another transformation will contain "real" number of classes.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 7 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1322]: IDE does not complain about this line. So It looks correct.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 10 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py]: Removed

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 11 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py]: Done

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 12 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1333]: Done.

**Review by rkazants — APPROVED** [review]

**Review by evolosen — APPROVED** [review]

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 9 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:1318]: This number cannot be deduced from the model. This parameter defines number of internal object that will be handled inside the DO operation. So it does not impact the output shapes of the DO operation. The number 6000 was selected rather arbitrary but verified with the customer dataset that it produces correct results on all images. I can hardly imagine that there could be more than 6000 detections per image, so this is a reasonable upper bound which could be applicable for any model.

**Review by mvafin — APPROVED** [review]
